### PR TITLE
Add Knapsack in C code

### PR DESCRIPTION
**Pull request by Zuoxun**
Running results are as follows:
>不超过背包容量的最大物品价值为 270
不超过背包容量的最大物品价值为 270       
不超过背包容量的最大物品价值为 270       
不超过背包容量的最大物品价值为 270

If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/c/chapter_dynamic_programming/knapsack.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/**
+ * File: knapsack.c
+ * Created Time: 2023-10-02
+ * Author: Zuoxun (845242523@qq.com)
+ */
+
+#include "../utils/common.h"
+
+int cap;
+int wgt_size;
```
Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:10]: https://github.com/krahets/hello-algo/pull/828#discussion_r1346165079

File: codes/c/chapter_dynamic_programming/knapsack.c
Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:1]: Please add code execution results

Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:1]: Missing cmake file

File: codes/c/chapter_dynamic_programming/knapsack.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    // 若超过背包容量，则只能不放入背包
+    if (wgt[i - 1] > c) {
+        return knapsackDFS(wgt, val, i - 1, c);
+    }
+    // 计算不放入和放入物品 i 的最大价值
+    int no = knapsackDFS(wgt, val, i - 1, c);
+    int yes = knapsackDFS(wgt, val, i - 1, c - wgt[i - 1]) + val[i - 1];
+    // 返回两种方案中价值更大的那一个
+    return max(no, yes);
```
Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:26]: The `max` function is not included in the c code public header file

Comment (Gonglja) [reply codes/c/chapter_dynamic_programming/knapsack.c:26]: The `max` function cannot be placed in vector.h

File: codes/c/chapter_dynamic_programming/knapsack.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    // 计算不放入和放入物品 i 的最大价值
+    int no = knapsackDFSMem(wgt, val, mem, i - 1, c);
+    int yes = knapsackDFSMem(wgt, val, mem, i - 1, c - wgt[i - 1]) + val[i - 1];
+    // 记录并返回两种方案中价值更大的那一个
+    mem[i][c] = max(no, yes);
+    return mem[i][c];
+}
+
+/* 0-1 背包：动态规划 */
+int knapsackDP(int wgt[], int val[], int cap) {
```
Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:52]: The `wgt` size is passed in as a parameter, or using vector.h

File: codes/c/chapter_dynamic_programming/knapsack.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                // 不选和选物品 i 这两种方案的较大值
+                dp[i][c] = max(dp[i - 1][c], dp[i - 1][c - wgt[i - 1]] + val[i - 1]);
+            }
+        }
+    }
+    return dp[n][cap];
+}
+
+/* 0-1 背包：空间优化后的动态规划 */
+int knapsackDPComp(int wgt[], int val[], int cap) {
```
Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:73]: Same as above

Comment (Gonglja) [inline codes/c/chapter_dynamic_programming/knapsack.c:73]: Same as above

File: codes/c/utils/vector.h
Comment (Gonglja) [inline codes/c/utils/vector.h:1]: https://github.com/krahets/hello-algo/pull/828#discussion_r1346171982

File: codes/c/chapter_dynamic_programming/knapsack.c
Comment (Zuoxun) [reply codes/c/chapter_dynamic_programming/knapsack.c:1]: Running results are as follows:
>不超过背包容量的最大物品价值为 270
不超过背包容量的最大物品价值为 270       
不超过背包容量的最大物品价值为 270       
不超过背包容量的最大物品价值为 270

**Review by Gonglja — COMMENTED** [review]
Approve @krahets

**Review by krahets — APPROVED** [review]

**Pull request merged**